### ISSUE TITLE
Anyway to import my own RIME input pattern?

### ISSUE THREAD

**Issue by sdugoten** [Cantoboard/Cantoboard#13]
I like the app, but I just don't know how to use the LSHK way of input.  I am using RIME on PC using similar input as Kaiboard.  Anyway if you can provide a way to choose the Kaiboard way of input, or better yet to allow people to import their own RIME config?  Thanks.

**Comment by alex-the-man** [comment]
Does Yale(耶魯拼音) mode work for you? You could switch to Yale(耶魯拼音) by long pressing the status indicator on the top right corner.

**Comment by alex-the-man** [comment]
Which Rime schema are you using? If you've customize it, could you share your config?
Import is on the road map. I will start working on it after polishing existing features.

**Comment by sdugoten** [comment]
[hkcantonese.dict.yaml.txt](https://github.com/Cantoboard/Cantoboard/files/6696913/hkcantonese.dict.yaml.txt)

This is the input pattern I use, which should match what Kaiboard user use

**Comment by alex-the-man** [comment]
The scheme is Yale + non standard romanization.
Listing the non standard romanization here so I could add it to Yale mode:
Could you try Yale mode and list all the finals you have difficulty with?

I will list a few. Yale standard form on the left, non standard form on the right:
1. e -> eh
2. ui -> eui
3. am -> um
4. eun -> un
5. uen -> en
6. u -> ue
7. un -> oon
8. ut -> oot
9. o -> oh
10.

**Comment by sdugoten** [comment]
ok, I will try to find some words that I have difficulty with

子  ji
算 suen
重 chung
春 chun
搬 boon
猜 chaai
測 chaak
慘  chaam
抄 chaau
齊  chai
尋  cham
扯 che
將  jeung
長  cheung
請  ching
千  chin
粗 cho
才  choi
創  chong
全 chuen
anything start with "ch"  like 速 隨 彩 差 ..etc
攰  guei
句  gui
向  heung
軟  yuen

some like that I guess?

**Comment by alex-the-man** [comment]
@sdugoten Let me add it, give me a couple days. Once it's released, would you be interested in testing the beta build on TestFlight? https://testflight.apple.com/join/zq9YSjuv

**Comment by alex-the-man** [comment]
> 猜 chaai
> 測 chaak
> 慘 chaam
> 抄 chaau
> 齊 chai
> 尋 cham
> 扯 che

Have you tried Yale mode?

**Comment by sdugoten** [comment]
of course I would like to test the beta when it is available.

using the Yale method, that cut down quite a few issue, but here are the rest

算 suen or any other word end with uen
估  goo
越  yuet
對 dui


One thing is that if the input method would allow to type english and chinese together?  For example, if I want to type

"佢英文名叫Lumd Pazk"

For kaiboard user, we will type "keui ying man meng giu Lumd Pazk <enter>"
For Cantoboard, I will have to type "keui ying man meng giu Lumd <enter> Pazk <enter>"

Basically, if you type in some english that is not regonize by the input method, you will have to press enter to show that word.

**Comment by alex-the-man** [comment]
> 算 suen or any other word end with uen
> 估 goo
> 越 yuet
> 對 dui

These are non standard forms which we will add soon.
Just wondering do you see Cantonese input method merely as a tool? Do you see additional value in it? Like correcting mispronunciation, reinforcing Cantonese, etc.

> Basically, if you type in some english that is not regonize by the input method, you will have to press enter to show that word.
This's by design, as Apple's builtin 拼音輸入法 behaves in that way. If you don't like it, I could add an option to change the behaviour.

**Comment by sdugoten** [comment]
> Just wondering do you see Cantonese input method merely as a tool? Do you see additional value in it? Like correcting mispronunciation, reinforcing Cantonese, etc.

I mostly use it as a tool.  Having said that, I do think other people might like the way you create this tool also help on correcting mispronuciation tho.  One thing I would like to add is that if it is possible to change the dollar sign $ default to something else as I am currently in UK, I mostly use £ sign.

> 
> This's by design, as Apple's builtin 拼音輸入法 behaves in that way. If you don't like it, I could add an option to change the behaviour.

that will be very helpful! thanks.

**Comment by alex-the-man** [comment]
> One thing I would like to add is that if it is possible to change the dollar sign $ default to something else as I am currently in UK, I mostly use £ sign.

Could you create another issue? Also, could you show me sign in Apple's builtin English keyboard?

**Comment by alex-the-man** [comment]
https://github.com/Cantoboard/Cantoboard/commit/3c366d26b654683d1959980b8b852e01617c8af0

**Comment by alex-the-man** [comment]
https://testflight.apple.com/join/zq9YSjuv

**Comment by alex-the-man** [comment]
Feel free to open a new issue if these rules are not sufficient.

**Comment by sdugoten** [comment]
So far the new rules works.  The only remaining part is this:

"佢英文名叫Lumd Pazk"

For kaiboard user, we will type "keui ying man meng giu Lumd Pazk [enter]"
For Cantoboard, I will have to type "keui ying man meng giu Lumd [enter] Pazk [enter]"

And I found two more issue

1. For example when you want type type the English word "luk", it will automatically change into "六" when you press space.  I think it's regarding the same algorithm on inputting English word.

2. Another thing is that when you type "Cha" and then space in Kaiboard, it will show as "Cha" as we expected.  If you change your mind and want to turn the word "Cha" into "差", you just need to backspace once, and the word list will show all the Chinese word pronounce as  "cha".  I think somehow Kaiboard will make a temp buffer to store the last typed word in case you need to change the English word back into Chinese.


Do you want me to raise a separate case for the 2 issue above?

**Comment by alex-the-man** [comment]
I think it's covered by https://github.com/Cantoboard/Cantoboard/issues/14.

Let me distill the behaviour of Kaiboard and post an update in that issue. Feel free to document the Kaiboard behaviour in that issue too.

**Comment by sdugoten** [comment]
some other word I run into problem is

豬, 住, which Yale spell that as "jyu" while most canton input spell it as "ju".  
書, 輸 which Yale spell as "syu" while most canton input spell it as "sue"

**Comment by alex-the-man** [comment]
Syu -> sue is supported and is working:
![image](https://user-images.githubusercontent.com/8400790/129651509-ce971027-79a0-4b74-b8d3-385f354783a6.png)

For jyu -> ju, I've to check how many ambiguous cases it will cause before adding the rule.

**Comment by alex-the-man** [comment]
I added yu -> u rule here: https://github.com/Cantoboard/Cantoboard/commit/62ebfe57286382a2632418bb08000cb1970097f9

**Comment by sdugoten** [comment]
> Syu -> sue is supported and is working:
> ![image](https://user-images.githubusercontent.com/8400790/129651509-ce971027-79a0-4b74-b8d3-385f354783a6.png)
> 

I didn't realize it's at end of the selection list.  It seems frequent keyword wouldn't put "書" in front of those English char.

> For jyu -> ju, I've to check how many ambiguous cases it will cause before adding the rule.

In fact, you don't have to force yourself to make this an official mapping.  I guess once you have a custom mapping rule that allow user to do it on their own, let user to do whatever they want and you can leave this as is.

I personally use RIME input on Mac, Windows and IOS and they all share the same input pattern so that it is always consistent.  I would just wait for custom mapping rules as you said you did have a plan for user to do their own rule

**Comment by alex-the-man** [comment]
Supporting custom rules will be really complicated due to iOS sandbox limitation of keyboard extension.

**Comment by sdugoten** [comment]
I now use cantoboard exclusively on a daily basis to test it out.  More on the mapping

woon > 換
goon > 觀
poon > 潘

The above spelling same issue as sue 書 which the chinese character selection is at the end of all english words.  And frequent words wouldn't put the word to the front

yee > 以

**Comment by alex-the-man** [comment]
> The above spelling same issue as sue 書 which the chinese character selection is at the end of all english words. And frequent words wouldn't put the word to the front

Cantoboard guesses the intended language (English or Chinese) from the input. Right now the prediction is really bad. I'm thinking of a way to improve it.

**Comment by alex-the-man** [comment]
oon -> un
ee -> i

**Comment by sdugoten** [comment]
算 〉suen
Arrow key you see above can be changed to english keyboard version when typing Chinese?

**Comment by alex-the-man** [comment]
I just realize that the following words are supported already. I guess you could see them if you expand the candidate pane. I'm working on improving the ordering.
woon > 換
goon > 觀
poon > 潘

Only yee > 以 needs a new rule.

> Arrow key you see above can be changed to english keyboard version when typing Chinese?
If you expand the candidate pane then press the status "耶" indicator, you could switch:

![image](https://user-images.githubusercontent.com/8400790/130312260-73b5554e-fb2d-4f1d-933c-95771d8eb0d5.png)

**Comment by alex-the-man** [comment]
https://github.com/Cantoboard/Cantoboard/commit/24aed52dc947ef671d854563da071c3769572dec

**Comment by alex-the-man** [comment]
I've published the build to TestFlight: https://testflight.apple.com/join/zq9YSjuv

**Comment by sdugoten** [comment]
One more
理 〉lee

Anyway to default the symbo to use"半＂ when typing Chinese ？

**Comment by alex-the-man** [comment]
> 理 〉lee

https://github.com/Cantoboard/Cantoboard/commit/5b2c71f776d4afaf70bad501707446fe15d86aa2

**Comment by alex-the-man** [comment]
> Anyway to default the symbo to use"半＂ when typing Chinese ？

![image](https://user-images.githubusercontent.com/8400790/130334721-ef145238-3b1f-41e6-affb-f4a3e7cd511e.png)

**Comment by alex-the-man** [comment]
> Also, when I try to type "半＂ ， there is a space after the word 半 and before the quote. Using Kaiboard can type without space "半"

I tried to reproduce it but I couldn't. Could you provide the instruction?

**Comment by sdugoten** [comment]
> > Also, when I try to type "半＂ ， there is a space after the word 半 and before the quote. Using Kaiboard can type without space "半"
> 
> I tried to reproduce it but I couldn't. Could you provide the instruction?

Here you go.

https://streamable.com/v887kc

You notice there is a missing space after the word "typing" and an extra space after the word 半.  The second part of the video showed that I did the same thing using Kaiboard which have no such behavior.

**Comment by alex-the-man** [comment]
The ending quote you entered is in full shape: '＂', not in half shape: '"'.
There are 2 ways to enter half shaped double quote:
- Changing the setting to use half shape as the default
![image](https://user-images.githubusercontent.com/8400790/130335974-d6fdf988-0681-41cf-ba1f-c1102e0c5aa1.png)
- Long press the double quote key, it will show all available double quote characters in different forms. 
![image](https://user-images.githubusercontent.com/8400790/130335995-5f6ed862-7d74-45e5-9947-65a436b89ac3.png)

**Comment by sdugoten** [comment]
I change the keyboard to"半" half shape and that solve part of the issue.  I can now type"半" without a space after the word 半, but the missing space before the first quote is still an issue.

**Comment by alex-the-man** [comment]
> but the missing space before the first quote is still an issue.

I reproduced this. Smart space removal incorrectly remove the space char before the starting quote. I will add some logic to fix this. I will create another issue.

**Comment by alex-the-man** [comment]
If you need more fuzzy rules, feel free to reopen this issue.
